### DISCUSSION TITLE
[Help] [Resolved] Combine multiple greyscale images - red channel not working correctly?

### DISCUSSION THREAD

**Discussion by matanshukry** [ImageMagick/ImageMagick#2704] [Help] [answered]
I have 4 greyscale images (ignore out.png for now):
![image](https://user-images.githubusercontent.com/5613042/95654757-d8371880-0b0a-11eb-8102-54ef4a6b485c.png)

and I want to create a single rgba image, with each one in a channel.
I tried doing this

magick red.png green.png blue.png alpha.png -colorspace rgb -channel RGBA -combine out.png

and while it mostly work, it seems there is a problem with the red channel:
![image](https://user-images.githubusercontent.com/5613042/95654753-d2413780-0b0a-11eb-92f5-6acc6131862b.png)

What am I missing here?

**Comment by snibgo** [comment]
What problem with the red channel? Your inputs red.png and green.png appear to be the same. When a pixel has RGB=(100%,100%,0), it is yellow.

**Comment by matanshukry** [comment]
@snibgo Sorry I played with it a lot got confused when did it again to post this issue.
I did it again, and here are the correct images, which shows the problem:

![image](https://user-images.githubusercontent.com/5613042/95657245-1d177b00-0b1c-11eb-8596-4b9edf5d76b4.png)

And the result:

![image](https://user-images.githubusercontent.com/5613042/95657298-67006100-0b1c-11eb-89df-4d7b0161cd18.png)

You can see the big green-ish cloud, which is the green channel.
However it seems it has red in it, all over it. As can be seen by the color (taken using the color picker), 168 red.

**Comment by snibgo** [comment]
Please post the actual input files.

Before the "-combine", you convert the inputs (assumed sRGB) to linear RGB with "-colorspace rgb". Why do you do this? If the inputs are black and white only, this shouldn't make any difference. But I suspect you don't really want to do this.

**Reply by matanshukry** [reply]
Regaring the colorspace parameter: I used it because at one point I added -verbose flag, without the colorspace parameter, and noticed it says the output is sRGB. I'm not 100% sure, but I believe I want rgb and not sRGB, which is why I added it.

According to your comment though I'm assuming it affects the input? I was assuming it affects the output when I added it actually.


Actual input files:

[files.zip](https://github.com/ImageMagick/ImageMagick/files/5359670/files.zip)

edit:
The original files I'm using are at 8k resolutoin and 16-bit greyscale images.
I noticed when uploading them they lose some of the precision, so I put them together in a zip with a password
to make sure they're intact.

The password is "1234".

**Comment by fmw42** [comment]
`convert red green blue -combine alpha -alpha off -compose copy_opacity -composite result`

**Reply by matanshukry** [reply]
Awesome, this works! however I am getting this warning at the end:
```
magick: profile 'icc': 'GRAY': Gray color space not permitted on RGB PNG `result' @ warning/png.c/MagickPNGWarningHandler/1748.
```

What does it mean exactly? any side effects that are "hidden" in the result?

**Reply by fmw42** [reply]
What was the command?

**Reply by matanshukry** [reply]
Pretty much what you wrote with the extra extensions to files:
```
magick red.png green.png blue.png -combine alpha.png -alpha off -compose copy_opacity -composite result.png
```

**Comment by snibgo** [comment]
> profile 'icc' ... What does it mean exactly?

It means your first input has an embedded ICC profile. "-verbose info:" says it is grayscale sRGB. So you need to convert the inputs to colour sRGB, or simply strip the profile.

red.png doesn't contain any white. Converting it to linear RGB will change values.

**Reply by matanshukry** [reply]
> So you need to convert the inputs to colour sRGB, or simply strip the profile.

How do I do any of them?

> red.png doesn't contain any white

What do you mean by that?
when I open the file with GIMP I can clearly see it says "Grayscale 16-bit gamma integer",
and using the picker tool it shows white on the area.

> Converting it to linear RGB will change values

How do I do that?

**Comment by snibgo** [comment]
```
f:\web\im>%IMG7%magick red.png -verbose info:
Image: red.png
  Format: PNG (Portable Network Graphics)
  Mime type: image/png
  Class: DirectClass
  Geometry: 8193x8193+0+0
  Resolution: 118.11x118.11
  Print size: 69.3676x69.3676
  Units: PixelsPerCentimeter
  Colorspace: Gray
  Type: Grayscale
  Base type: Undefined
  Endianess: Undefined
  Depth: 16-bit
  Channel depth:
    Gray: 16-bit
  Channel statistics:
    Pixels: 67125249
    Gray:
      min: 0  (0)
      max: 60992 (0.930678)
      mean: 3524.58 (0.0537817)
```
The maximum value on red.png is 0.930678, on a scale of 0.0 to 1.0. So there are no white pixels.

To convert your inputs to colour sRGB, use "-profile sRGB.icc". If that icc file isn't in your current directory, include the full path.

To strip the profile, use "+profile *".

**Reply by matanshukry** [reply]
What happens if I don't, other than the warning?

To me it seems intuitive that magick would just take the value and make it red; so 0.93 would be 0.93 red.
I'm guesing something else happens though? why then? and what exactly?

**Comment by fmw42** [comment]
ImageMagick tries to use the first image in the command line as a template in terms of colorspace and type for the output. In your case, the first image is grayscale with a gray icc profile (icc:description: GIMP built-in D65 Grayscale with sRGB TRC), but you want the output to be color (3 channels with alpha after combining). So ImageMagick and PNG object to assigning a grayscale profile copied from the first input to a color image. So it is giving that message, which is just a warning and not an error.  You can probably prevent the message by adding -quiet to your command right after convert.

**Comment by snibgo** [comment, marked as answer]
Yes, it is just a warning. IM will create the colour output PNG with embedded grayscale ICC profile. But this is not a good idea, because following software won't know what to do with the profile.

**Reply by fmw42** [reply]
Good point. 

One needs to remove the profile and add an sRGB.icc or similar color profile at the end. Probably add -colorspace sRGB to the command

(untested)
`magick red.png green.png blue.png -set colorspace sRGB -combine +profile "*" -profile path_to/sRGB.icc alpha.png -alpha off -compose copy_opacity -composite result.png`

**Reply by matanshukry** [reply]
Cool great.

And thank you both for all the help and all the information!

**Comment by westjensontexas** [comment]
Thanks for the clarification.
